### PR TITLE
Fix ldc opcode in internal contexts

### PR DESCRIPTION
**Pull request by Dentosal**
LDC opcode was behaving incorrectly in internal contexts, i.e. when called from within a contract.

https://github.com/FuelLabs/fuel-specs/pull/584

## Checklist
- [x] Breaking changes are clearly marked as such in the PR description and changelog
- [x] New behavior is reflected in tests

### Before requesting review
- [x] I have reviewed the code myself

### CODE REVIEW

File: fuel-vm/src/interpreter/blockchain/code_tests.rs [hunk 1]
```diff
@@ -14,7 +14,7 @@ use crate::{
 use fuel_tx::Contract;
 
 #[test]
-fn test_load_contract() -> IoResult<(), Infallible> {
+fn test_load_contract_in_script() -> IoResult<(), Infallible> {
```
Comment (MitchTurner) [inline fuel-vm/src/interpreter/blockchain/code_tests.rs:17]: Shouldn't these tests be against the `Interpreter` instead of against `LoadContractCodeCtx`?

Comment (Dentosal) [reply fuel-vm/src/interpreter/blockchain/code_tests.rs:17]: They should probably be against both. I still hate that `LoadContractCodeCtx` and the others like that even exist.

**Review by xgreenx — DISMISSED** [review]

File: fuel-vm/src/interpreter/blockchain.rs [hunk 2]
```diff
@@ -636,19 +636,20 @@ where
...
-            let old_code_size = Word::from_be_bytes(self.memory.read_bytes(fp)?);
-
+        // Update frame code size, if we have a stack frame (i.e. fp > 0)
+        if self.context.is_internal() {
+            let code_size_ptr =
+                (*self.fp).saturating_add(CallFrame::code_size_offset() as Word);
+            let old_code_size =
+                Word::from_be_bytes(self.memory.read_bytes(code_size_ptr)?);
+            let old_code_size = padded_len_word(old_code_size)
+                .expect("Code size cannot overflow with padding");
```
Comment (Dentosal) [inline fuel-vm/src/interpreter/blockchain.rs:646]: I would assume the size should already be padded when written in memory. Is it really correct to just compute padding here?

Comment (xgreenx) [reply fuel-vm/src/interpreter/blockchain.rs:646]: Our code uses unpadded value in the `CallFrame`

image

**Review by xgreenx — APPROVED** [review]

File: fuel-vm/src/interpreter/blockchain.rs (hunk 2 above)
Comment (Dentosal) [reply fuel-vm/src/interpreter/blockchain.rs:646]: Yep. Shouldn't `CallFrame` use the padded value?

Comment (xgreenx) [reply fuel-vm/src/interpreter/blockchain.rs:646]: I don't know=) The question is to the compiler team: do they use it somehow or not? Or is it used by anyone=)

Our specification says "Code size in bytes.", that sounds like exact size of the contract without padding.

image

Comment (Dentosal) [reply fuel-vm/src/interpreter/blockchain.rs:646]: I'll ask the team, but a quick search from Sway codebase seems like they don't use this for anything. Time to remove? :D

Comment (Voxelot) [reply fuel-vm/src/interpreter/blockchain.rs:646]: Doesnt "word-aligned" mean the same thing as padding here?

Comment (Dentosal) [reply fuel-vm/src/interpreter/blockchain.rs:646]: Yep.

Moved to use padded size in 7335f9fe.

**Comment by xgreenx** [comment]
Could ou also update the Specification to specify this behaviour explicitly, please?=)

File: fuel-vm/src/tests/blockchain.rs [hunk 3]
```diff
@@ -2047,3 +2047,180 @@ fn coinbase_works() {
 
     assert_eq!(data.as_ref().unwrap(), &*expected);
 }
+
+#[test]
+fn various_ldc_issues_poc() {
```
Comment (MitchTurner) [inline fuel-vm/src/tests/blockchain.rs:2052]: Is there any way to break this down in to more granular && specific tests? Tests like this are hard to maintain as it's not clear what behavior they are defining.

File: fuel-vm/src/interpreter/blockchain/code_tests.rs (hunk 1 above)
Comment (MitchTurner) [reply fuel-vm/src/interpreter/blockchain/code_tests.rs:17]: Okay. I don't you to fix the architecture in this PR.

File: fuel-vm/src/interpreter/flow.rs [hunk 4]
```diff
@@ -579,27 +584,35 @@ where
...
+        mem_frame.copy_from_slice(&frame.to_bytes());
+        let (mem_code, mem_code_padding) = mem_code.split_at_mut(code_size);
+        read_contract(call.to(), self.storage, mem_code)?;
+        mem_code_padding.fill(0);
+
+        #[allow(clippy::arithmetic_side_effects)] // Checked above
+        let code_start =
+            (*self.registers.system_registers.fp) + CallFrame::serialized_size() as Word;
+
+        *self.registers.system_registers.pc = code_start;
```
[the remark this answers is not in the document]
Comment (Dentosal) [reply fuel-vm/src/interpreter/flow.rs:603]: No? Essentially it was before like this:

```python
content_size = len - frame_len_with_padding
frame_end = fp + content_size # return value
```

[the remark this answers is not in the document]
Comment (xgreenx) [reply fuel-vm/src/interpreter/flow.rs:603]: Ohh, there is `saturating_sub`, I missed that=D

**Review by xgreenx — APPROVED** [review]

**Pull request merged**